feat : make dynamic graph title(#3)

Rengod95 opened the pull request:
**작업 이슈**
- https://github.com/ChartStudio/Dynamic_Chart/issues/3
- edit controller's init
- add title property on Options Interface, BasicConfig and Styler class
- edit Animation class
- add drawCanvasTitle function on GraphView
- add dependencies
- edit test.html BaseConfig

**작업 상세 설명**

- test.html 의 BaseConfig로 title 값을 받을 수 있도록 변경
- controller와 animation의 렌더링 로직에 title 렌더링 로직 추가 

**참조 문서**
- N/A

**추가 변경 사항**

refactor : make dynamic graph title(https://github.com/ChartStudio/Dynamic_Chart/issues/3)

- BaseConfig 의 title property 삭제
- BaseConfig 내에 chart_title property추가

- ChartTitleConfig 추가
  - ChartTitleConfig 클래스 추가
  - ChartTitle Interface 추가
  - TitleType 추가
  - /type/index 수정
  - /config/index 수정

Rengod95 committed:
feat : make dynamic graph title(#3)

- edit controller's init
- add title property on Options Interface, BasicConfig and Styler class
- edit Animation class
- add drawCanvasTitle function on GraphView
- add dependencies
- edit test.html BaseConfig

Rengod95 committed:
feat : make dynamic graph title(#3)

- styler method 오타 수정

Rengod95 committed:
Merge pull request #9 from ChartStudio/ISSUE-3

feat : make dynamic graph title(#3)

MK-Lee13 commented:
```
- https://github.com/ChartStudio/Dynamic_Chart/issues/3
```

작업 이슈에 다음과 같이 issue url 을 입력하면 아래와 같이 이슈를 표시할 수 있으며,
해당 이슈에 PR이 걸리게 됩니다. 참고하면 좋을 것 같습니다.

***작업 이슈***
- https://github.com/ChartStudio/Dynamic_Chart/issues/3

MK-Lee13 commented on package.json:
```diff
@@ -33,5 +33,10 @@
         "ts-node": "^10.9.1",
         "typescript": "^4.9.4"
+    },
+    "dependencies": {
```
해당 의존성은 지워주세요. 현재 해당 프로젝트에 필요한 의존성이 아닙니다.
단순 개발 의존성에 필요한 내용입니다.
```
  "live": "^0.1.25-beta.0",
  "server": "^1.0.38",
  "tslib": "^2.4.1"
```

MK-Lee13 replied:
`"dependencies"` 통으로 지워주시면 됩니다.

MK-Lee13 replied:
해당 작업 이후 npm install 하셔서 package-lock.json 원복하시면 될 것 같습니다.

Rengod95 replied:
확인했습니다. 의존성 수정 후 npm install 진행하였습니다!

MK-Lee13 commented on src/view/view.graph.ts:
```diff
@@ -37,6 +37,11 @@ class GraphView {
 
+  drawCanvasTitle(){
+    let pixel = this.position.getRectPixel();
+    this.context?.fillText(this.styler.getGraphTitle(),  pixel.x/2, 20)
```
```
this.context?.fillText(this.styler.getGraphTitle(), pixel.x / 2, 20)
```
현재 eslint 가 없어서 컨벤션 체크하는 것이 의미가 있는 것은 아니지만 수식의 경우 뛰어쓰는 것을 선호합니다.

MK-Lee13 commented on src/core/core.styler.ts:
```diff
@@ -12,18 +13,19 @@ class Styler {
 
   constructor(context: CanvasRenderingContext2D | null, config: BaseConfig) {
     this.context = context;
+    this.title = config.title;
```
현재 작업 내용의 경우 title 에 대한 value 값만을 받아서 처리하고 있습니다.
해당 경우 라이브러리 활용면에서 좋은 효과를 주지 못합니다.

css를 예시로 들었을 때 우리가 어떠한 글을 쓸때
- 폰트
- 색상
- 해당 태그 내부에 border 스타일을 입힐지
등등 많은 것을 고려할 수가 있습니다.

라이브러리 개발 시에도 해당 부분을 고려하면 좋을 것 같습니다.
즉 현재
```
this.title = config.title # 이런 형식 보다는
this.titleConfig = config.titleConfig # 이런 형식으로
```

단순 Config 내용만 변경하더라도 커스터 마이징이 가능한 형식으로 작업해주시면 좋을 것 같습니다.

MK-Lee13 reviewed (commented):
작업 수고하셨습니다~~👍
리뷰 남긴 부분 작업 참고하여 작업 후 Approve 사인 받으시면 그때 머징하시면 될 것 같습니다.

Rengod95 committed:
refactor : make dynamic graph title(#3)

- BaseConfig 의 title property 삭제
- BaseConfig 내에 chart_title property추가

- ChartTitleConfig 추가
  - ChartTitleConfig 클래스 추가
  - ChartTitle Interface 추가
  - TitleType 추가
  - /type/index 수정
  - /config/index 수정

Rengod95 committed:
Merge pull request #10 from ChartStudio/ISSUE-3

refactor : make dynamic graph title(#3)

MK-Lee13 commented on package-lock.json:
```diff
@@ -1,531 +1,1289 @@
 {
-  "name": "Dynamic_Chart",
```
현재 package-lock.json 이 차이가 발생하고 있습니다. 아마 저와 노드 버전 이 달라서 차이가 나오는 것으로 확인됩니다.

package-lock.json 이 수시로 변경될 위험이 있기에 노드 버전 을 동기화 해야할 것 같습니다.
현재 제가 사용중인 노드 버전 의 경우 `18.12.1` 을 사용하고 있습니다. 노드 버전 동기화 시켜주시고, package-lock.json 에 대한 변경 사항을 없애주셔야 합니다.

혹시 타 프로젝트로 인해 노드 버전 변경이 어려운경우 코멘트 달아주시면 감사하겠습니다.

MK-Lee13 commented on src/type/type.chart.title.ts:
```diff
@@ -0,0 +1,9 @@
+
+export interface ChartTitle{
+    type : TitleType// 제목의 타입 설정
+    content : string, // 제목의 내용
```
기본 값이 들어가게 전부 변경해야 합니다. 이럴 경우 전부다 nullable 한 형식으로 되어야 합니다.
차트 타이틀은 사용자가 입력할 수도 있고 없을 수도 있기에 해당 부분 고려해서 nullable 로 변경해주세요.

MK-Lee13 commented on src/type/type.chart.title.ts:
```diff
@@ -0,0 +1,9 @@
+export interface ChartTitle{
+    type : TitleType// 제목의 타입 설정
+    content : string, // 제목의 내용
+    options?: any,
```
any 부분 제거 부탁드립니다. 구조화가 가능한 파트로 보이는데 any 를 사용할 이유가 없습니다.

MK-Lee13 commented on src/config/config.chart.title.ts:
```diff
@@ -0,0 +1,43 @@
+
+class ChartTitleConfig{
+    private content: string;
+    private options?: any;
```
동일하게 any 제거 부탁드립니다. any 를 사용하는 것은 정말 어쩔 수 없는 경우에만 사용하셔야 합니다.
만약 그렇지 않다면 저희가 type script 를 사용할 이유는 없습니다.

사용가 전달하는 정보를 정규화하기 위해서 현재 타입이 있는 언어를 선택하여 사용중에 있습니다.

MK-Lee13 commented on src/core/core.config.ts:
```diff
@@ -1,8 +1,16 @@
 
 interface Options {
   type: string;
+  chart_title: ChartTitle;
```
option 의 경우 `chart_title` 보다 `title` 이 조금 더 적합하다고 생각합니다.

MK-Lee13 commented on src/core/core.config.ts:
```diff
@@ -23,15 +31,27 @@ interface Options {
+  type: "standard",
+  content: 'default chart title',
+  options : {},
+}
```
해당 부분 사용하지 않는 데이터로 보입니다. 삭제 부탁드립니다.

MK-Lee13 commented on src/core/core.config.ts:
```diff
@@ -47,6 +67,7 @@ class BaseConfig {
   verticalConfig: VerticalStyleConfig;
   backgroundConfig: BackgroundStyleConfig;
   lineConfigList: LineStyleConfig[];
+  chart_title:ChartTitleConfig;
```
js 컨벤션의 경우 `snake` 가 아닌 `camelCase` 를 사용하는 것으로 알고 있습니다.
컨벤션이 중요하지 않다고 생각하실수 있지만 규약이기에 지켜주시면 감사하겠습니다. 그리고 타입 한칸만 뛰어주시면 감사하겠습니다.

MK-Lee13 commented on src/core/core.controller.ts:
```diff
@@ -56,18 +56,20 @@ class BaseChart {
-    this.canvas.style.width = config.width + 'px';
-    this.canvas.style.height = config.height + 'px';
+    this.canvas.style.width = config.width+ 100 + 'px';
+    this.canvas.style.height = config.height + 100 + 'px';
```
다음과 같이 `100` 이라는 숫자가 필요할 경우 default 값으로 상수로 선언하여 위로 빼주세요. 

또한 해당 부분은 조건부로 빠져야 합니다. 타이틀 작성을 희망하지 않는 사용자가 있을 수 도 있기 때문에 해당 부분 고려해서 코드 작성해 주시면 감사하겠습니다.

MK-Lee13 commented on src/core/core.styler.ts:
```diff
@@ -1,29 +1,38 @@
 
 class Styler {
   private context: CanvasRenderingContext2D | null;
+  private chart_title : ChartTitleConfig;
```
위 커멘트에 동일내용 적었습니다. 컨벤션 신경써주시면 감사하겠습니다.
